Integrate wporg custom warnings

amieiro opened the pull request:
The GlotPress installed at [WordPress.org](https://translate.wordpress.org/) has [a plugin](https://meta.svn.wordpress.org/sites/trunk/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php) that implements some custom translation warnings. It [adds some new warnings and makes changes](https://github.com/WordPress/wordpress.org/blob/78ec5a0d77a9dc18698779b26d0efb9c6135190f/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php) to some core Glotpress warnings.

This PR migrates those warnings from the custom plugin to GlotPress.

The warnings are managed in GlotPress in two classes, both in the "[gp-includes/warnings.php](https://github.com/GlotPress/GlotPress-WP/blob/e96b261665151e47e451eb2cce854133ee48e63b/gp-includes/warnings.php)" file:

* [GP_Translation_Warnings](https://github.com/GlotPress/GlotPress-WP/blob/e96b261665151e47e451eb2cce854133ee48e63b/gp-includes/warnings.php#L14), used to handle translation warnings.
* [GP_Builtin_Translation_Warnings](https://github.com/GlotPress/GlotPress-WP/blob/e96b261665151e47e451eb2cce854133ee48e63b/gp-includes/warnings.php#L124),  used to register built-in translation warnings. The common warnings are defined in the methods of this class.

This PR:

1. Adds a new CLI command (**GP_CLI_Wporg** class, available at "gp-includes/cli/wporg.php") to be able to set the installation as the wporg installation. This CLI command sets or removes an option variable (gp_is_wporg) in the options table. The usage of this command is very simple:
    * To set the installation as the wporg installation, execute ```wp glotpress wporg set```
    * To remove the installation as the wporg installation, execute ```wp glotpress wporg unset```
2. Adds or updates the methods from the wporg plugin that can fit in the generic solution in the [GP_Builtin_Translation_Warnings](https://github.com/GlotPress/GlotPress-WP/blob/e96b261665151e47e451eb2cce854133ee48e63b/gp-includes/warnings.php#L124) class:
    * Adds the **warning_mismatching_urls()** method. This method adds a warning for changing plain-text URLs. This method allows:
         * Change the scheme (https <=> http). 
         * Include/remove a trailing slash in the URL.
         * For some domains to change to a subdomain. 

         This method has a hook to update the list of domains with allowed changes (empty by default). More info in the point 3.

    * Adds the **warning_unexpected_sprintf_token()** method. This method adds a warning for adding unexpected percent signs in a sprintf-like string. This is to catch translations for originals like this:
	    *  Original: ```<a href="%s">100 percent</a>```
	    *  Submitted translation: ```<a href="%s">100%</a>```
	    *  Proper translation: ```<a href="%s">100%%</a>```

    * Updates the **warning_tags()** method with the wporg check rules:
        *  URLs (href + src) are run through the previous method (warning_mismatching_urls()).
	        * The domain may change for some safe domains.
	        * The protocol may change between https & http.
	        * The URL may include/remove a trailing slash.
	    * The value of translatable/url attributes is excluded from the error message if it's not related to the issue at hand.
	    *  Tags are sorted, ```<em>One</em> <strong>Two</strong>``` can be translated as ```<strong>foo</strong> <em>bar</em>``` without generating warning.
	    *  East Asian languages can remove emphasis/italic tags. 

    * Adds two new rules to the **warning_tags()** method, not present in the wporg plugin:
        * 1st rule checks if the HTML tags are in correct order. This one warns about HTML tags translations in incorrect order. For example:
            * Original: ```<a></a>```
            * Translation: ```</a><a>```
        * 2nd rule checks whether links that are not URL or placeholders are equal or not. For example:
            * Original: ```<a href="%s" title="Blimp!">Baba</a>```
            * Translation: ```<a href="javascript:%s" title="Блимп!">Баба</a>```

3. Adds a new class (**GP_Wporg_Translation_Warnings**, available at "gp-includes/wporg/warnings.php") to handle wporg installation specific elements that do not fit in the generic plugin. Currently, this class has two relevant methods:
    * **add_wporg_allowed_domains()** adds the allowed domains in the GP_Builtin_Translation_Warnings class using the _gp_allowed_domain_changes_ hook present in the warning_mismatching_urls() method.  The current domains with allowed changes to own subdomains are:
        * wordpress.org
        * wordpress.com
        * en.gravatar.com
        * en.wikipedia.org
    * **gp_add_all_warnings()** method adds all warning methods from this class in the GP_Builtin_Translation_Warnings class using the _gp_add_all_warnings_ hook in the add_all() method. 
        
    Currently, this new class has only a new warning method, **warning_wporg_mismatching_placeholders()**, that adds a warning for changing placeholders. This method only supports placeholders in the format of '###[A-Z_]+###'. This warning method is so specific that only fits in the wporg installation.
    The two new hooks present in the GP_Builtin_Translation_Warnings class (_gp_allowed_domain_changes_ and _gp_add_all_warnings_) are only "hooked" if a variable (gp_is_wporg) is set in the options table, setting the installation as the wporg installation.

    Note: This PR includes the changes of [this other PR](https://github.com/GlotPress/GlotPress-WP/pull/1237).

## Architecture 

This architecture, which separates the common warnings from the wporg warnings and injects them via hooks, is extensible, because:

* If the plugin needs a new warning for all the installations, it can be placed in the GP_Builtin_Translation_Warnings class and its tests in the GP_Test_Builtin_Translation_Warnings class.
* If the warning only fits for the wporg installation, it can be placed in the GP_Wporg_Translation_Warnings (and its tests in the GP_Test_Wporg_Warnings class) and inject the code using a current hook (or a new one if it would be required) in the GP_Builtin_Translation_Warnings class. 
* This approach allows to eliminate the need for a custom plugin, because all its business logic is injected using hooks, and it is easier to customize with new warnings. 
* This approach allows to keep all the business logic of the wporg installation in its own classes and folders, isolated from the main plugin. This allows not to run the custom code for the wporg installation on a normal installation, which will not have the option table variable (gp_is_wporg) enabled by default.
        
## Tests

I have added new tests for these new or updated methods. I have divided the tests in two files:
* **GP_Test_Builtin_Translation_Warnings** class (tests/phpunit/testcases/test_builtin_warnings.php file). This is an existent class. I have created or updated the test methods to adapt it to the new logic. These tests are for the common functionality.
* **GP_Test_Wporg_Warnings** class (tests/phpunit/testcases/test_wporg/test_wporg_warnings.php file). This is a new class to test the new methods that will only be executed in the wporg installation.

### Test improvements

* This PR adds a new method (**assertHasWarningsAndContainsOutput()**) in both test classes that replaces and extends the functionality of the _assertHasWarnings()_ method. This method is responsible for checking the same as the _assertHasWarnings()_ method and also checks the text returned by each warning, so we can assure the warning output text.
* This PR adds a new method (**test_chained_warnings()** in the GP_Test_Builtin_Translation_Warnings class) that checks that several chained warnings show the corresponding errors. This is a kind of integration test, where the different warning methods are tested working together.

amieiro committed:
Adds a warning for changing plain-text URLs

This method allows for the scheme to change, and for some domains
to change to a subdomain.

This method integrates a warning from an existing plugin at
https://translate.wordpress.org , available at
https://github.com/WordPress/wordpress.org/blob/master/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L37 ,
so that this validation is available to all GlotPress users.

amieiro committed:
Updates the GlotPress tags warning to allow some URL changes.

This method integrates a warning from an existing plugin at
https://translate.wordpress.org , available at
https://github.com/WordPress/wordpress.org/blob/master/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L131 ,
so that this validation is available to all GlotPress users.

- URLs (href + src) are run through `self::warning_mismatching_urls()`
    - The domain may change for some safe domains
    - The protocol may change between https & http
    - The URL may include/remove a trailing slash
- The value of translatable/URL attributes is excluded from the error message if it's not related to the issue at hand.
- Tags are validated to be nested correctly, validating the ordering of the tags remained the same.
- East Asian languages can remove emphasis/italic tags.

amieiro committed:
Adds a warning for changing placeholders.

This only supports placeholders in the format of '###[A-Z_]+###'.
This method integrates a warning from an existing plugin at
https://translate.wordpress.org ,
available at https://github.com/WordPress/wordpress.org/blob/master/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L238 ,
so that this validation is available to all GlotPress users.

amieiro committed:
Adds a warning for adding unexpected percent signs in a sprintf-like string.

This is to catch translations for originals like this:
 - Original: `<a href="%s">100 percent</a>`
 - Submitted translation: `<a href="%s">100%</a>`
 - Proper translation: `<a href="%s">100%%</a>`

 This method integrates a warning from an existing plugin at
    https://translate.wordpress.org ,
    available at https://github.com/WordPress/wordpress.org/blob/master/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L276 ,
    so that this validation is available to all GlotPress users.

amieiro committed:
Corrects some placeholder warnings

This code comes from the PR https://github.com/GlotPress/GlotPress-WP/pull/1237
and integrates an update in a filter from an existing plugin at
https://translate.wordpress.org ,
available at https://github.com/WordPress/wordpress.org/blob/7c67af3e7b49333ce3c5d565e57fa439c9c52223/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L312 ,
so that this check is available to all GlotPress users.

amieiro committed:
Rename the AssertHasWarnings to assertHasWarnings

amieiro committed:
Adds some tests to the placeholder warnings

Adds a new test to increase the code coverage and other tests to check the
output when the check has warnings

amieiro committed:
Adds some tests to check the output of newline validation methods

amieiro committed:
Adds some tests to check the output of warning_mismatching_urls() method

amieiro committed:
Adds some tests to the check() method

Adds two new tests to increase the code coverage

amieiro committed:
Adds a new test to the warning_length() method

Adds a new test to an edge case, to increase the code coverage

amieiro committed:
Add a CLI command that sets/unsets a wporg variable in the options table

This option value will be used to conditionally execute code related with the wporg installation

amieiro committed:
Separate the specific warnings for wporg

Create a new class (GP_Wporg_Translation_Warnings) for the wporg warnings
and inject it using hooks in the GP_Builtin_Translation_Warnings class.

Inject the specific wporg list of domains with allowed changes to their own
subdomains and a specific wporg warning for changing placeholders in the
format of '###[A-Z_]+###'.

amieiro committed:
Remove some validations and add one

- Remove the tag order validation, because this depends of the translation language
- Remove the validation that checks if the links that are not URL are equal or not
- Check if the translation tags are in correct order. This method warns about HTML
  tags translations in incorrect order. For example:
  - Original: <a></a>
  - Translation: </a><a>

amieiro committed:
Check in the HTML tags whether links that are not URL or placeholders are equal or not

amieiro committed:
Add the PhpDocumentor header in the GP_CLI_Wporg class

amieiro committed:
Add a todo in the GP_Wporg_Translation_Warnings class

amieiro committed:
Refactor the test code

Include the assertContainsOutput() and the assertHasWarnings() methods in one
new method with the same functionality that both methods.

Refactor the GP_Test_Builtin_Translation_Warnings class to comply with the
WordPress Coding Standards.

amieiro committed:
Add a bugfix from the wporg current plugin

See:
- https://wordpress.slack.com/archives/C0341VCNL/p1623107393007500
- https://github.com/WordPress/wordpress.org/commit/9032284195ae6044dad655690fe483e9ad4eb230#diff-f2a4e047f07ebba7056d82966599302848df1e6f4418ca3e40dfbaca46ef3c06
- https://github.com/WordPress/wordpress.org/commit/36901bfc27e865b8ceb0fa7d63ae1d0bdf0c0ddb#diff-f2a4e047f07ebba7056d82966599302848df1e6f4418ca3e40dfbaca46ef3c06

amieiro committed:
Test that several chained warnings show the corresponding errors

ocean90 commented:
Thanks for working on this! What was your motivation to adopt the custom warnings?

Before going into a deep review, everything related to `GP_Wporg_Translation_Warnings` and the CLI command should be removed again as this doesn't belong to the plugin. There's nothing wrong with still using a plugin on wordpress.org which can customize the warnings.

ocean90 commented:
Looking at the two specific warnings, I think they can be integrated too without the need to set a `wporg` flag. The domains should only be extendable which they are thanks to the filter and we'd a good name for `warning_wporg_mismatching_placeholders`, maybe `warning_hashtag_placeholders` or `warning_named_placeholders`?

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -201,49 +213,122 @@ public function warning_length( $original, $translation, $locale ) {
 	 * @return string|true True if check is OK, otherwise warning message.
 	 */
-	public function warning_tags( $original, $translation, $locale ) {
+	function warning_tags( $original, $translation, $locale ) {
```
```suggestion
	public function warning_tags( $original, $translation, $locale ) {
```

amieiro replied:
Done

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -153,6 +153,16 @@ class GP_Builtin_Translation_Warnings {
+	 *
+	 * @var array
+	 */
+	public $allowed_domain_changes = array();
```
This should probably contain the domains from `GP_Wporg_Translation_Warnings` as default.

amieiro replied:
Done.

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -401,8 +620,123 @@ function ( $key ) {
+	 * Returns the values from the href and the src
+	 *
+	 * @since 3.0.0
+	 * @access public
```
```suggestion
	 * @access private
```

Can you review the `@access` tags again? There are a few others which don't match the actually visibility.

To be honest, I'm fine if you just keep them out.

amieiro replied:
I have updated the visibility to two functions, to maintain the consistency with the other functions, that have the "@access" parameter

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -401,8 +620,123 @@ function ( $key ) {
+				$message[] = trim( $error->message );
+			}
+			return sprintf(
+				__( 'The translation tags are not correct: %s', 'glotpress' ),
```
```suggestion
				__( 'The translation contains incorrect HTML tags: %s', 'glotpress' ),
```

This makes the message consistent with others and clarifies that this is about HTML tags .

amieiro replied:
Done.

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -388,6 +473,140 @@ public function warning_should_not_end_on_newline( $original, $translation, $loc
+	 *
+	 * @param string $original    The original string.
+	 * @param string $translation The translated string.
+	 * @return string|true        True if check is OK, otherwise warning message.
```
```suggestion
	 * @return string|true True if check is OK, otherwise warning message.
```

`@return` and `@param` tags don't get aligned.

amieiro replied:
Done in two methods.

ocean90 commented on gp-includes/warnings.php:
```diff
@@ -201,49 +213,122 @@ public function warning_length( $original, $translation, $locale ) {
+		// East asian languages can remove emphasis/italic tags
+		if ( count( $original_parts ) > count( $translation_parts ) ) {
+
+			$languages_without_italics = array(
+				'ja',
+				'ko',
+				'zh',
+				'zh-hk',
+				'zh-cn',
+				'zh-sg',
+				'zh-tw',
+			);
```
Can we move this to a class property like `$length_exclude_languages`?

amieiro replied:
Done.

amieiro committed:
Remove the CLI command for the wporg

amieiro committed:
Move items from the wporg warning parametrization to the warning core

Remove the wporg parametrization
Move and adapt the tests from the wporg parametrization to the warning core tests
Add 4 allowed domains

amieiro committed:
Add the visibility to the warning_tags() method

amieiro committed:
Remove two incorrect alignments in the @return parameter of two methods

amieiro commented:
I have removed all the wporg parametrization, including the CLI command, and I have integrated the domains and the specific warning in the builtin warnings. I have named the moved method as "warning_mismatching_placeholders", to [maintain the same name that it has now in the wporg plugin](https://github.com/WordPress/wordpress.org/blob/3527c98b2d9f69e25907ab92d764ed3ae8d1a525/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L238).

amieiro committed:
Add translator comments in the strings with placeholders

amieiro committed:
Small changes to adapt it to the code standard

ocean90 commented:
> I have named the moved method as "warning_mismatching_placeholders", to [maintain the same name that it has now in the wporg plugin](https://github.com/WordPress/wordpress.org/blob/3527c98b2d9f69e25907ab92d764ed3ae8d1a525/wordpress.org/public_html/wp-content/plugins/wporg-gp-custom-warnings/wporg-gp-custom-warnings.php#L238).

I still don't think that this is a good name when you compare it with `warning_placeholders`. Just from reading the name I'd expect that both are about the same placeholders. But one is about PHP placeholders and the new one about "named" placeholders. Maybe we should consider renaming `warning_placeholders` to `warning_php_placeholders` but that's something for a separate issue. With `warning_mismatching_placeholders` it would also get complicated to find another name if we plan to support another type in the future. What do you think?

amieiro commented:
I think we can change ```warning_mismatching_placeholders``` to ```warning_translation_placeholders``` so that whoever sees it understands that this method checks translation placeholders. In another PR we could change ```warning_placeholders``` to ```warning_php_placeholders```. Do I make the change?

amieiro committed:
Rename the warning_mismatching_placeholders() method to warning_translation_placeholders()

amieiro committed:
Rename the warning_translation_placeholders() method to warning_named_placeholders()

amieiro committed:
Merge branch 'GlotPress:develop' into integrate-wporg-gp-custom-warnings

amieiro committed:
Update changelog
